File: packages/vault-core/src/hlc.ts
/**
 * Hybrid Logical Clock.
 *
 * Last-write-wins ordered by `Date.now()` is itself a data-loss bug: a device
 * whose clock runs slow can never win a merge, so its edits are silently
 * discarded forever. An HLC keeps wall-clock readability but guarantees that
 * an edit which *causally follows* another always sorts after it, whatever the
 * two devices' clocks say.
 *
 * Encoded as a lexicographically sortable string so it can be compared with
 * `<` and stored in JSON without a custom comparator:
 *
 *   000001723456789-00003-a1b2c3d4
 *   \_____________/ \___/ \______/
 *      wall ms     counter  node
 */

const WALL_DIGITS = 15;
const COUNTER_DIGITS = 5;
const MAX_COUNTER = 10 ** COUNTER_DIGITS - 1;

/** A point in causal time. Compare these with `<`, `>`, or `compareHlc`. */
export type Hlc = string;

export type HlcParts = { wall: number; counter: number; node: string };

export function encodeHlc(parts: HlcParts): Hlc {
  const wall = String(parts.wall).padStart(WALL_DIGITS, "0");
  const counter = String(parts.counter).padStart(COUNTER_DIGITS, "0");
  return `${wall}-${counter}-${parts.node}`;
}

export function decodeHlc(value: Hlc): HlcParts {
  const wall = value.slice(0, WALL_DIGITS);
  const counter = value.slice(WALL_DIGITS + 1, WALL_DIGITS + 1 + COUNTER_DIGITS);
  const node = value.slice(WALL_DIGITS + COUNTER_DIGITS + 2);
  const parsedWall = Number.parseInt(wall, 10);
  const parsedCounter = Number.parseInt(counter, 10);
  if (Number.isNaN(parsedWall) || Number.isNaN(parsedCounter) || node.length === 0) {
    throw new Error(`Malformed HLC timestamp: ${value}`);
  }
  return { wall: parsedWall, counter: parsedCounter, node };
}

export function compareHlc(a: Hlc, b: Hlc): number {
  return a < b ? -1 : a > b ? 1 : 0;
}

/** The zero timestamp — sorts before every real one. Used for absent values. */
export const HLC_ZERO: Hlc = encodeHlc({ wall: 0, counter: 0, node: "" });

export interface Clock {
  /** Stamp a new local event. */
  now(): Hlc;
  /** Observe a remote timestamp, advancing local causal time past it. */
  observe(remote: Hlc): void;
  /** Current state, for persistence across restarts. */
  snapshot(): Hlc;
}

export function createClock(options: {
  node: string;
  physical?: () => number;
  /** Restore causal time after a restart so a fast local clock can't be undone. */
  resume?: Hlc | undefined;
}): Clock {
  const physical = options.physical ?? (() => Date.now());
  if (options.node.length === 0) throw new Error("An HLC needs a non-empty node id.");

  let last: HlcParts = options.resume
    ? { ...decodeHlc(options.resume), node: options.node }
    : { wall: 0, counter: 0, node: options.node };

  function advance(candidateWall: number, candidateCounter: number): Hlc {
    if (candidateCounter > MAX_COUNTER) {
      // Counter saturation means >100k events in one millisecond. Borrow from
      // the next millisecond rather than wrapping, which would break ordering.
      last = { wall: candidateWall + 1, counter: 0, node: options.node };
    } else {
      last = { wall: candidateWall, counter: candidateCounter, node: options.node };
    }
    return encodeHlc(last);
  }

  return {
    now() {
      const wall = physical();
      if (wall > last.wall) return advance(wall, 0);
      return advance(last.wall, last.counter + 1);
    },
    observe(remote: Hlc) {
      const parts = decodeHlc(remote);
      const wall = physical();
      const maxWall = Math.max(wall, last.wall, parts.wall);
      let counter: number;
      if (maxWall === last.wall && maxWall === parts.wall) {
        counter = Math.max(last.counter, parts.counter) + 1;
      } else if (maxWall === last.wall) {
        counter = last.counter + 1;
      } else if (maxWall === parts.wall) {
        counter = parts.counter + 1;
      } else {
        counter = 0;
      }
      advance(maxWall, counter);
    },
    snapshot() {
      return encodeHlc(last);
    },
  };
}

File: packages/android/vault/src/main/kotlin/app/keyweb/vault/Hlc.kt
package app.keyweb.vault

/**
 * Hybrid Logical Clock.
 *
 * Last-write-wins ordered by wall time is itself a data-loss bug: a device
 * whose clock runs slow can never win a merge, so its edits are silently
 * discarded forever. An HLC keeps wall-clock readability but guarantees that an
 * edit which causally follows another always sorts after it, whatever the two
 * devices' clocks say.
 *
 * Encoded as a lexicographically sortable string, identical to the TypeScript
 * implementation so both platforms read each other's vaults:
 *
 *     000001723456789-00003-a1b2c3d4
 *     \_____________/ \___/ \______/
 *        wall ms     counter  node
 */
private const val WALL_DIGITS = 15
private const val COUNTER_DIGITS = 5
private const val MAX_COUNTER = 99_999

/** A point in causal time. Ordering is plain string ordering. */
typealias Hlc = String

data class HlcParts(val wall: Long, val counter: Int, val node: String)

fun encodeHlc(parts: HlcParts): Hlc {
    val wall = parts.wall.toString().padStart(WALL_DIGITS, '0')
    val counter = parts.counter.toString().padStart(COUNTER_DIGITS, '0')
    return "$wall-$counter-${parts.node}"
}

fun decodeHlc(value: Hlc): HlcParts {
    require(value.length > WALL_DIGITS + COUNTER_DIGITS + 2) { "Malformed HLC timestamp: $value" }
    val wall = value.substring(0, WALL_DIGITS).toLongOrNull()
    val counter = value.substring(WALL_DIGITS + 1, WALL_DIGITS + 1 + COUNTER_DIGITS).toIntOrNull()
    val node = value.substring(WALL_DIGITS + COUNTER_DIGITS + 2)
    require(wall != null && counter != null && node.isNotEmpty()) {
        "Malformed HLC timestamp: $value"
    }
    return HlcParts(wall, counter, node)
}

/** The zero timestamp; sorts before every real one. Used for absent values. */
val HLC_ZERO: Hlc = encodeHlc(HlcParts(0, 0, ""))

/**
 * Stamps local events and absorbs remote ones.
 *
 * Not thread safe by itself: [VaultSync] holds it behind the same mutex that
 * serialises syncs, and edits go through that same path.
 */
class Clock(
    private val node: String,
    private val physical: () -> Long = { System.currentTimeMillis() },
    resume: Hlc? = null,
) {
    init {
        require(node.isNotEmpty()) { "An HLC needs a non-empty node id." }
    }

    private var last: HlcParts =
        resume?.let { decodeHlc(it).copy(node = node) } ?: HlcParts(0, 0, node)

    private fun advance(candidateWall: Long, candidateCounter: Int): Hlc {
        last = if (candidateCounter > MAX_COUNTER) {
            // Counter saturation means >100k events in one millisecond. Borrow
            // from the next millisecond rather than wrapping, which would break
            // ordering.
            HlcParts(candidateWall + 1, 0, node)
        } else {
            HlcParts(candidateWall, candidateCounter, node)
        }
        return encodeHlc(last)
    }

    /** Stamp a new local event. */
    fun now(): Hlc {
        val wall = physical()
        return if (wall > last.wall) advance(wall, 0) else advance(last.wall, last.counter + 1)
    }

    /** Observe a remote timestamp, advancing local causal time past it. */
    fun observe(remote: Hlc) {
        val parts = decodeHlc(remote)
        val wall = physical()
        val maxWall = maxOf(wall, last.wall, parts.wall)
        val counter = when {
            maxWall == last.wall && maxWall == parts.wall -> maxOf(last.counter, parts.counter) + 1
            maxWall == last.wall -> last.counter + 1
            maxWall == parts.wall -> parts.counter + 1
            else -> 0
        }
        advance(maxWall, counter)
    }

    /** Current state, for persistence across restarts. */
    fun snapshot(): Hlc = encodeHlc(last)
}
